buidl/psbt.py

```diff
@@ -541,6 +546,194 @@ def serialize(self):
+                # TODO: add support for legacy TXs?
+                raise SuspiciousTransaction(
+                    f"Input #{cnt} does not contain a witness script. "
+                    "This tool can only sign p2wsh transactions."
+                )
+
+            # for calculating msig fingerprint
+            root_xfp_hexes = []
+            for _, details in psbt_in.named_pubs.items():
+                root_xfp_hexes.append(details.root_fingerprint.hex())
+
+            quorum_m, quorum_n = psbt_in.witness_script.get_quorum()
+            if quorum_n != len(root_xfp_hexes):
+                raise SuspiciousTransaction(
+                    f"Transaction claims {quorum_m}-of-{quorum_n} but has {len(root_xfp_hexes)} root fingerprints, which != {quorum_n}"
+                )
+
+            input_desc = {
+                "quorum": f"{quorum_m}-of-{len(root_xfp_hexes)}",
+                "root_xfp_hexes": root_xfp_hexes,
+                "prev_txhash": psbt_in.tx_in.prev_tx.hex(),
+                "prev_idx": psbt_in.tx_in.prev_index,
+                "n_sequence": psbt_in.tx_in.sequence,
+                "sats": psbt_in.tx_in.value(),
+                "addr": psbt_in.witness_script.address(testnet=self.testnet),
+                # if adding support for p2sh in the future, the address would be: psbt_in.witness_script.p2sh_address(testnet=self.testnet),
+                "witness_script": str(psbt_in.witness_script),
+                "msig_id": calc_multisig_id(
+                    quorum_m=quorum_m, root_xfp_hexes=root_xfp_hexes
+                ),
+            }
+            inputs_desc.append(input_desc)
+
+        if not all(x["msig_id"] == inputs_desc[0]["msig_id"] for x in inputs_desc):
+            raise SuspiciousTransaction(
+                "Multiple different multisig wallets used in inputs. "
+                "Construct a transaction with one input to use this feature."
+            )
+
+        TOTAL_INPUT_SATS = sum([x["sats"] for x in inputs_desc])
+
+        # This too only supports TXs with 1-2 outputs (sweep TX OR spend+change TX):
+        if len(self.psbt_outs) > 2:
+            raise SuspiciousTransaction(
+                f"This tool does not support batching, your transaction has {len(self.psbt_outs)} outputs. "
+                "Please construct a transaction with <= 2 outputs."
+            )
+
+        spend_addr, output_spend_sats = "", 0
+        change_addr, output_change_sats = "", 0
+        outputs_desc = []
+        for cnt, psbt_out in enumerate(self.psbt_outs):
+            psbt_out.validate()
+
+            output_desc = {
+                "sats": psbt_out.tx_out.amount,
+                "addr": psbt_out.tx_out.script_pubkey.address(testnet=self.testnet),
+                "addr_type": psbt_out.tx_out.script_pubkey.__class__.__name__.rstrip(
+                    "ScriptPubKey"
+                ),
+            }
+
+            if psbt_out.named_pubs:
```

Comment:
I can see a case where you're naming multiple outputs, one which goes to one of the pubkeys of the quorum (maybe paying out a bet). I'm not sure this is a safe assumption.

Reply:
Can you describe this a little more? I'm wondering if I should pull this code block out of this pull review and save it for another one.

I want developers to be able to rely on battle-tested logic for building apps around. I'm trying to keep this simple, but it's a little unwieldy.

Reply:
You're assuming here that if the output is a bunch of named public keys (that is, known to the signers) that it's a change output. This may not be the case. If, for example you have a 2-of-2 output being spent where some amount of it goes back to the 2-of-2 but the rest goes to one of the parties (say in a bet that pays out multiple times depending on the conditions), then the spending output will be caught as a change output according to this logic. Does that make sense?

Reply:
Hmm, I think I might be misunderstanding what you're saying :(

In order to be counted as change, the current code requires that the output shares the same `xfp`s and `m-of-n` as every input`*`:
https://github.com/buidl-bitcoin/buidl-python/blob/55e09f551e6e40f6dd238b443bbe585c8be145a3/buidl/psbt.py#L635-L656

Shifting from an input controlled by the same group to a new output controlled by that same group does indeed sound like change to me (and the behavior that an end-user manually verifying would expect).

What am I missing? I don't want to introduce a vulnerability here!

`*` I am a little nervous relying on the `PSBT` class, but the logic in `validate()` seems solid as far as confirming that the xpubs in the PSBT correspond to the ones being used in the actual transactions. However, the xfp hex is the only part that can't be verified (it would be trivial to swap this out). I think in my `calc_multisig_id` method I should be using the xpubs (which are validated by the `PSBT` class) and not the xfps (which can only be validated by each Signer who holds the corresponding xpriv)!